In [1]:
!pip install pandas
!pip install numpy


  Using cached pytz-2025.2-py2.py3-none-any.whl.metadata (22 kB)
  Using cached tzdata-2025.2-py2.py3-none-any.whl.metadata (1.4 kB)
   ---------------------------------------- 0.0/11.0 MB ? eta -:--:--
   ------------------ --------------------- 5.0/11.0 MB 27.4 MB/s eta 0:00:01
   ------------------------------------ --- 10.0/11.0 MB 25.9 MB/s eta 0:00:01
   ---------------------------------------- 11.0/11.0 MB 25.4 MB/s eta 0:00:00
Using cached pytz-2025.2-py2.py3-none-any.whl (509 kB)
Using cached tzdata-2025.2-py2.py3-none-any.whl (347 kB)

   ---------------------------------------- 0/3 [pytz]
   ---------------------------------------- 0/3 [pytz]
   ------------- -------------------------- 1/3 [tzdata]
   ------------- -------------------------- 1/3 [tzdata]
   -------------------------- ------------- 2/3 [pandas]
   -------------------------- ------------- 2/3 [pandas]
   -------------------------- ------------- 2/3 [pandas]
   -------------------------- ------------- 2/3 [pand

In [2]:
import pandas as pd
import os
import numpy as np

## Reading the Data

In [3]:
flights= pd.read_csv("C:\\flight_price_segmaker\\Data\\flight_price.csv")

In [81]:
PROJECT_DIR =r"C:\flight_price_segmaker\Data"

In [31]:
def get_data(name):
    file_name=f"{name}.csv"
    file_path = os.path.join(PROJECT_DIR,file_name)
    return pd.read_csv(file)

In [4]:
flights.head()

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897  
1    05:50         13:15   7h 25m     2 stops         No info   7662  
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882  
3    18:05         23:30   5h 25m      1 stop         No info   6218  
4    16:50         21:35   4h 45m      1 stop         No info  13302

In [5]:
flights.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10683 non-null  object
 1   Date_of_Journey  10683 non-null  object
 2   Source           10683 non-null  object
 3   Destination      10683 non-null  object
 4   Route            10682 non-null  object
 5   Dep_Time         10683 non-null  object
 6   Arrival_Time     10683 non-null  object
 7   Duration         10683 non-null  object
 8   Total_Stops      10682 non-null  object
 9   Additional_Info  10683 non-null  object
 10  Price            10683 non-null  int64 
dtypes: int64(1), object(10)
memory usage: 918.2+ KB


- The dataset contains 10,683 roww and 11 features
- The data types of some features are not appropriate(date)

##### Observation
- The datatyepe of 'Data_of_journey','Dep_Time','Arrival_Time' should be datetime
- The tper of 'Duration' is mixed.It should be numeric type
- There are 220 duplicates these to be removed

In [6]:
flights.duplicated().sum()

220

In [7]:
(
    flights.loc[flights.duplicated(keep=False)]
            .sort_values(["Airline","Date_of_Journey","Source","Destination"])
)

Airline Date_of_Journey    Source Destination                  Route  \
6321  Air India      01/03/2019  Banglore   New Delhi  BLR → BOM → AMD → DEL   
9848  Air India      01/03/2019  Banglore   New Delhi  BLR → BOM → AMD → DEL   
572   Air India      03/03/2019  Banglore   New Delhi              BLR → DEL   
8168  Air India      03/03/2019  Banglore   New Delhi              BLR → DEL   
1495  Air India       1/04/2019   Kolkata    Banglore  CCU → DEL → COK → BLR   
...         ...             ...       ...         ...                    ...   
2692   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
2870   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
3711   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
2634    Vistara      24/03/2019  Banglore   New Delhi              BLR → DEL   
5470    Vistara      24/03/2019  Banglore   New Delhi              BLR → DEL   

     Dep_Time  Arrival_Time Duration Total_Stops  \
6321    08:50  23:55 02 Mar   39h 5m     2 stops   
9848    08:50  23:55 02 Mar   39h 5m     2 stops   
572     21:10         23:55   2h 45m    non-stop   
8168    21:10         23:55   2h 45m    non-stop   
1495    10:00  01:20 02 Apr  15h 20m     2 stops   
...       ...           ...      ...         ...   
2692    05:45         08:35   2h 50m    non-stop   
2870    05:45         08:35   2h 50m    non-stop   
3711    20:30         23:20   2h 50m    non-stop   
2634    11:30         14:10   2h 40m    non-stop   
5470    11:30         14:10   2h 40m    non-stop   

                   Additional_Info  Price  
6321                       No info  17135  
9848                       No info  17135  
572                        No info   7591  
8168                       No info   7591  
1495                       No info  10408  
...                            ...    ...  
2692  No check-in baggage included   4273  
2870  No check-in baggage included   4273  
3711  No check-in baggage included   3873  
2634                       No info   5403  
5470                       No info   5403  

[415 rows x 11 columns]

#### Detailed Analysis

In [8]:
flights.Airline.unique()

array(['IndiGo', 'Air India', 'Jet Airways', 'SpiceJet',
       'Multiple carriers', 'GoAir', 'Vistara', 'Air Asia',
       'Vistara Premium economy', 'Jet Airways Business',
       'Multiple carriers Premium economy', 'Trujet'], dtype=object)

In [9]:
(
    flights
    .Airline
    .str.replace("Premium economy","")
    .str.replace("Business","")
    .str.title()
    .unique()
)

array(['Indigo', 'Air India', 'Jet Airways', 'Spicejet',
       'Multiple Carriers', 'Goair', 'Vistara', 'Air Asia', 'Vistara ',
       'Jet Airways ', 'Multiple Carriers ', 'Trujet'], dtype=object)

- Some of the entries have inconsistence/inaccurate values

### Date_of_journey

In [10]:
flights.Date_of_Journey

0        24/03/2019
1         1/05/2019
2         9/06/2019
3        12/05/2019
4        01/03/2019
            ...    
10678     9/04/2019
10679    27/04/2019
10680    27/04/2019
10681    01/03/2019
10682     9/05/2019
Name: Date_of_Journey, Length: 10683, dtype: object

In [11]:
pd.to_datetime(flights.Date_of_Journey,dayfirst = True)

0       2019-03-24
1       2019-05-01
2       2019-06-09
3       2019-05-12
4       2019-03-01
           ...    
10678   2019-04-09
10679   2019-04-27
10680   2019-04-27
10681   2019-03-01
10682   2019-05-09
Name: Date_of_Journey, Length: 10683, dtype: datetime64[ns]

### Source

In [12]:
flights.Source.unique()

array(['Banglore', 'Kolkata', 'Delhi', 'Chennai', 'Mumbai'], dtype=object)

### Destination

In [13]:
flights.Destination.unique()

array(['New Delhi', 'Banglore', 'Cochin', 'Kolkata', 'Delhi', 'Hyderabad'],
      dtype=object)

### Dep_Time

In [14]:
flights.Dep_Time

0        22:20
1        05:50
2        09:25
3        18:05
4        16:50
         ...  
10678    19:55
10679    20:45
10680    08:20
10681    11:30
10682    10:55
Name: Dep_Time, Length: 10683, dtype: object

In [15]:
(
    flights
    .Dep_Time
    .loc[lambda ser:ser.str.contains("[^0-9:]")]  #tells wheather the col contain same format i.e 0-9: "^" is used as not which are not in that format are returned here it is none
)

Series([], Name: Dep_Time, dtype: object)

In [16]:
pd.to_datetime(flights.Dep_Time).dt.time

C:\Users\vishw\AppData\Local\Temp\ipykernel_18824\1056054517.py:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  pd.to_datetime(flights.Dep_Time).dt.time


0        22:20:00
1        05:50:00
2        09:25:00
3        18:05:00
4        16:50:00
           ...   
10678    19:55:00
10679    20:45:00
10680    08:20:00
10681    11:30:00
10682    10:55:00
Name: Dep_Time, Length: 10683, dtype: object

### Arrival_Time

In [17]:
flights.Arrival_Time

0        01:10 22 Mar
1               13:15
2        04:25 10 Jun
3               23:30
4               21:35
             ...     
10678           22:25
10679           23:20
10680           11:20
10681           14:10
10682           19:15
Name: Arrival_Time, Length: 10683, dtype: object

In [18]:
(
    flights
    .Arrival_Time
    .loc[lambda ser:ser.str.contains("[^0-9:]")]
    .str.split(" ",n=1)
    .str.get(1)
    .unique()
)

array(['22 Mar', '10 Jun', '13 Mar', '02 Mar', '10 May', '04 Mar',
       '13 Jun', '28 May', '19 Mar', '07 May', '02 Jun', '16 Jun',
       '19 May', '16 May', '28 Jun', '02 May', '28 Mar', '19 Jun',
       '04 Apr', '25 Mar', '07 Mar', '25 Jun', '07 Jun', '25 May',
       '13 May', '16 Mar', '22 May', '10 Apr', '04 Jun', '20 May',
       '28 Apr', '25 Apr', '10 Mar', '19 Apr', '13 Apr', '02 Apr',
       '23 Mar', '22 Apr', '11 May', '07 Apr', '03 May', '08 Mar',
       '03 Mar', '05 Mar', '22 Jun', '04 May', '26 May', '16 Apr',
       '26 Jun', '29 May', '29 Jun', '29 Mar', '23 May', '17 Jun'],
      dtype=object)

### Duration

In [21]:
flights.Duration

0        2h 50m
1        7h 25m
2           19h
3        5h 25m
4        4h 45m
          ...  
10678    2h 30m
10679    2h 35m
10680        3h
10681    2h 40m
10682    8h 20m
Name: Duration, Length: 10683, dtype: object

In [22]:
(
    flights
    .Duration
    .loc[lambda ser:~ser.str.contains("m")]
    .unique()  #"~" return which donot contain "m"
)

array(['19h', '23h', '22h', '12h', '3h', '5h', '10h', '18h', '24h', '15h',
       '16h', '8h', '14h', '20h', '13h', '11h', '9h', '27h', '26h', '4h',
       '7h', '30h', '21h', '28h', '47h', '6h', '25h', '38h', '34h'],
      dtype=object)

In [24]:
(
    flights
    .Duration
    .loc[lambda ser:~ser.str.contains("h")]
    # .unique()  #"~" return which donot contain "m"
)

6474    5m
Name: Duration, dtype: object

In [31]:
flights.iloc[[6474]]

Airline Date_of_Journey  Source Destination                  Route  \
6474  Air India       6/03/2019  Mumbai   Hyderabad  BOM → GOI → PNQ → HYD   

     Dep_Time Arrival_Time Duration Total_Stops Additional_Info  Price  
6474    16:50        16:55       5m     2 stops         No info  17327

In [44]:
(
    flights
    .Duration                                          
    .drop(index=6474)                                
    .str.split(" ",expand=True) #split at first space 2h 32m #if expand is true each value will become a col in dataframe format
    .set_axis(["hour","minute"],axis=1)
    .assign(
        hour=lambda df_:(
            df_
            .hour
            .str.replace("h","")
            .astype(int)
            .mul(60)
        ),
        minute = lambda df_:(
            df_
            .minute
            .str.replace("m","")
            .fillna("0")
            .astype(int)
        )
    )
    .sum(axis=1)
)  

0         170
1         445
2        1140
3         325
4         285
         ... 
10678     150
10679     155
10680     180
10681     160
10682     500
Length: 10682, dtype: int64

In [45]:
def convert_to_minutes(ser):
    return(
        ser
        .str.split(" ",expand=True) #split at first space 2h 32m #if expand is true each value will become a col in dataframe format
        .set_axis(["hour","minute"],axis=1)
        .assign(
            hour=lambda df_:(
                df_
                .hour
                .str.replace("h","")
                .astype(int)
                .mul(60)
            ),
            minute = lambda df_:(
                df_
                .minute
                .str.replace("m","")
                .fillna("0")
                .astype(int)
            )
        )
        .sum(axis=1)
    )

### Total_stops

In [51]:
flights.Total_Stops.unique()

array(['non-stop', '2 stops', '1 stop', '3 stops', nan, '4 stops'],
      dtype=object)

In [63]:
(
    flights
    .Total_Stops
    .replace("non-stop","0")
    .str.replace("stops?","",regex=True) # "stops" "stop" ? infront of s means s is opitonal
    .pipe(lambda ser:pd.to_numeric(ser))
    
)

0        0.0
1        2.0
2        2.0
3        1.0
4        1.0
        ... 
10678    0.0
10679    0.0
10680    0.0
10681    0.0
10682    2.0
Name: Total_Stops, Length: 10683, dtype: float64

### Additional_Info

In [69]:
flights.Additional_Info.unique()

array(['No info', 'In-flight meal not included',
       'No check-in baggage included', '1 Short layover', 'No Info',
       '1 Long layover', 'Change airports', 'Business class',
       'Red-eye flight', '2 Long layover'], dtype=object)

#### Cleaning Operations

In [70]:
def clean_data(df):
    return(
        df
        .drop(index=[6474])
        .assign(**{
            col: df[col].str.lower()
            for col in df.select_dtypes(include='O').columns
        })
        .rename(columns=str.lower)
        .assign(
            airline = lambda df_:(
                df_
                .airline
                .str.replace("Premium economy","")
                .str.replace("Business","")
                .str.title()
            ),
            date_of_journey = lambda df_:pd.to_datetime(df_.date_of_journey,dayfirst=True),
            dep_time = lambda df_: pd.to_datetime(df_.dep_time).dt.time,
            arrival_time = lambda df_: pd.to_datetime(df_.arrival_time).dt.time,
            duration = lambda df_: df_.duration.pipe(convert_to_minutes),
            total_stops = lambda df_:(
                df_
                .total_stops
                .replace("non-stop","0")
                .str.replace("stops?","",regex=True) # "stops" "stop" ? infront of s means s is opitonal
                .pipe(lambda ser:pd.to_numeric(ser))
            ),
            additional_info = lambda df_:df_.additional_info.replace("No info","No Info")
        )
        .drop(columns="route")
    )

In [73]:
flights_cleaned = clean_data(flights)
flights_cleaned

C:\Users\vishw\AppData\Local\Temp\ipykernel_18824\3748362066.py:19: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  dep_time = lambda df_: pd.to_datetime(df_.dep_time).dt.time,
C:\Users\vishw\AppData\Local\Temp\ipykernel_18824\3748362066.py:20: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  arrival_time = lambda df_: pd.to_datetime(df_.arrival_time).dt.time,


airline date_of_journey    source destination  dep_time  \
0           Indigo      2019-03-24  banglore   new delhi  22:20:00   
1        Air India      2019-05-01   kolkata    banglore  05:50:00   
2      Jet Airways      2019-06-09     delhi      cochin  09:25:00   
3           Indigo      2019-05-12   kolkata    banglore  18:05:00   
4           Indigo      2019-03-01  banglore   new delhi  16:50:00   
...            ...             ...       ...         ...       ...   
10678     Air Asia      2019-04-09   kolkata    banglore  19:55:00   
10679    Air India      2019-04-27   kolkata    banglore  20:45:00   
10680  Jet Airways      2019-04-27  banglore       delhi  08:20:00   
10681      Vistara      2019-03-01  banglore   new delhi  11:30:00   
10682    Air India      2019-05-09     delhi      cochin  10:55:00   

      arrival_time  duration  total_stops additional_info  price  
0         01:10:00       170          0.0         no info   3897  
1         13:15:00       445          2.0         no info   7662  
2         04:25:00      1140          2.0         no info  13882  
3         23:30:00       325          1.0         no info   6218  
4         21:35:00       285          1.0         no info  13302  
...            ...       ...          ...             ...    ...  
10678     22:25:00       150          0.0         no info   4107  
10679     23:20:00       155          0.0         no info   4145  
10680     11:20:00       180          0.0         no info   7229  
10681     14:10:00       160          0.0         no info  12648  
10682     19:15:00       500          2.0         no info  11753  

[10682 rows x 10 columns]

### Split the Data

In [74]:
flights_final = flights_cleaned.sample(1000)

In [75]:
X = flights_final.drop(columns="price")
y = flights_final.price.copy()

In [76]:
!pip install scikit-learn

  Using cached joblib-1.5.1-py3-none-any.whl.metadata (5.6 kB)
  Using cached threadpoolctl-3.6.0-py3-none-any.whl.metadata (13 kB)
   ---------------------------------------- 0.0/10.7 MB ? eta -:--:--
   ------------------------- -------------- 6.8/10.7 MB 38.3 MB/s eta 0:00:01
   ---------------------------------------- 10.7/10.7 MB 26.7 MB/s eta 0:00:00
Using cached joblib-1.5.1-py3-none-any.whl (307 kB)
   ---------------------------------------- 0.0/38.4 MB ? eta -:--:--
   ---- ----------------------------------- 3.9/38.4 MB 21.3 MB/s eta 0:00:02
   ---------- ----------------------------- 10.2/38.4 MB 25.5 MB/s eta 0:00:02
   -------------- ------------------------- 14.2/38.4 MB 24.0 MB/s eta 0:00:02
   ----------------- ---------------------- 17.0/38.4 MB 20.3 MB/s eta 0:00:02
   -------------------- ------------------- 19.9/38.4 MB 19.4 MB/s eta 0:00:01
   ------------------------ --------------- 23.6/38.4 MB 18.7 MB/s eta 0:00:01
   ---------------------------- ----------- 27

In [77]:
from sklearn.model_selection import train_test_split
X_,X_test,y_,y_test = train_test_split(X,y,test_size=0.2,random_state=42)
X_train,X_val,y_train,y_val = train_test_split(X_,y_,test_size=0.2,random_state=42)


In [78]:
print(X_train.shape, y_train.shape)
print(X_val.shape, y_val.shape)
print(X_test.shape, y_test.shape)

(640, 9) (640,)
(160, 9) (160,)
(200, 9) (200,)


### Export the subsets

In [82]:
def export_data(X, y, name):
	file_name = f"{name}.csv"
	file_path = os.path.join(PROJECT_DIR, file_name)

	X.join(y).to_csv(file_path, index=False)

	return pd.read_csv(file_path).head()

In [83]:
export_data(X_train,y_train,"train")

airline date_of_journey    source destination  dep_time arrival_time  \
0  Jet Airways      2019-06-12  banglore       delhi  17:45:00     20:45:00   
1  Jet Airways      2019-05-21     delhi      cochin  21:50:00     12:35:00   
2       Indigo      2019-04-24     delhi      cochin  20:25:00     01:30:00   
3        Goair      2019-04-21     delhi      cochin  10:35:00     19:35:00   
4  Jet Airways      2019-06-06     delhi      cochin  15:00:00     19:00:00   

   duration  total_stops              additional_info  price  
0       180          0.0                      no info   8016  
1       885          1.0  in-flight meal not included  12898  
2       305          1.0                      no info   5065  
3       540          1.0                      no info   7031  
4      1680          1.0  in-flight meal not included  10262

In [84]:
export_data(X_val, y_val, "val")

airline date_of_journey    source destination  dep_time arrival_time  \
0  Jet Airways      2019-06-24     delhi      cochin  18:15:00     12:35:00   
1    Air India      2019-03-21     delhi      cochin  06:05:00     09:25:00   
2  Jet Airways      2019-05-24  banglore       delhi  19:50:00     22:50:00   
3  Jet Airways      2019-03-18    mumbai   hyderabad  19:35:00     21:05:00   
4       Indigo      2019-04-15   kolkata    banglore  09:35:00     12:20:00   

   duration  total_stops              additional_info  price  
0      1100          1.0  in-flight meal not included  10262  
1      1640          1.0                      no info   7532  
2       180          0.0  in-flight meal not included   6144  
3        90          0.0  in-flight meal not included   2071  
4       165          0.0                      no info   4174

In [85]:
export_data(X_test, y_test, "test")

airline date_of_journey    source destination  dep_time arrival_time  \
0       Indigo      2019-05-15   kolkata    banglore  15:30:00     18:05:00   
1  Jet Airways      2019-06-01     delhi      cochin  21:25:00     12:35:00   
2     Air Asia      2019-03-03   kolkata    banglore  19:35:00     22:05:00   
3  Jet Airways      2019-05-09     delhi      cochin  05:30:00     19:00:00   
4    Air India      2019-03-01  banglore   new delhi  08:50:00     10:35:00   

   duration  total_stops additional_info  price  
0       155          0.0         no info   4804  
1       910          2.0         no info  13292  
2       150          0.0         no info   5964  
3       810          2.0         no info  15129  
4      1545          2.0         no info  31783